File: binary_tree/scala/ObjectOriented.scala
class Node[a <% Ordered[a] ](var value: a) {
    var left: Node[a] = null
    var right: Node[a] = null
    def insertInto(node: Node[a], value: a) {
        if(value < node.value) {
            if(node.left == null) node.left = new Node(value)
            else insertInto(node.left, value)
        } else {
            if(node.right == null) node.right = new Node(value)
          else insertInto(node.right, value)
        }
    }
}

class Tree[a <% Ordered[a]] extends Traversable[a] {
    var root: Node[a] = null
    def <<(value: a) {
        if(root == null) root = new Node(value)
        else root insertInto (root, value)
    }

    override def foreach[U](block: a => U) {
        getValueOn(root, block)
    }

    private def getValueOn(node: Node[a], block: a => Any) {
        if(node.left != null) getValueOn(node.left, block)
        block(node.value)
        if(node.right != null) getValueOn(node.right, block)
    }
}

object ObjectOriented extends App {
    val tree = new Tree[Int]
    tree << 10
    tree << 1
    tree << 20
    tree << 2
    tree << 2
    tree << 5
    tree << 3
    tree << 12
    tree << 7
    tree << 6
    tree << 9
    tree << 11
    tree << 8
    
    tree foreach println
    println(tree.toList)
}

File: binary_tree/ruby/object_oriented.rb
class Node
  attr_accessor :left, :right, :value
  protected :left=, :right=, :value=

  def initialize(value)
    @value = value
  end

  def insert_into(node, value)
    if value < node.value
      if node.left.nil?
        node.left = Node.new(value)
      else
        insert_into node.left, value
      end
    else
      if node.right.nil?
        node.right = Node.new(value)
      else
        insert_into node.right, value
      end
    end
  end
end

class Tree
  include Enumerable
  def <<(value)
    if @root.nil?
      @root = Node.new(value)
    else
      @root.insert_into @root, value
    end
  end

  def each(&b)
    get_value_on @root, &b
  end

  def get_value_on(node, &b)
    get_value_on node.left, &b if node.left
    b.call node.value
    get_value_on node.right, &b if node.right
  end
  private :get_value_on
end

tree = Tree.new
tree << 10
tree << 1
tree << 20
tree << 2
tree << 2
tree << 5
tree << 3
tree << 12
tree << 7
tree << 6
tree << 9
tree << 11
tree << 8

tree.each do |value|
  puts value
end

p tree.to_a
